lightkurve/correctors/metrics.py

```diff
@@ -0,0 +1,333 @@
+            # noise is greater than the uncertainties.
+            # So, when Sigmoid > 0.5 (given twiceSigmoidInv defn.)
+            result = np.sum(pgChange[pgChange > 0.0]) / (
+                (len(np.nonzero(pgChange > 0.0)[0])) * meanCorrectedUncertPower
+            )
+            metric_per_iter.append(result)
+
+    metric = np.mean(metric_per_iter)
+
+    # We want the goodness to span (0,1]
+    # Use twice a reversed sigmoid to get a [0,1] range mapped from a [0,inf) range
+    def sigmoidInv(x):
+        return 2.0 / (1 + np.exp(x))
+
+    # Make sure maximum score is 1.0
+    metric = sigmoidInv(np.max([metric, 0.0]))
+
+    return metric
+
+
+def underfit_metric_neighbors(
+    corrected_lc: LightCurve,
+    radius: float = 6000,
+    min_targets: int = 30,
+    max_targets: int = 50,
+    interpolate: bool = False,
+):
+    """This goodness metric measures the degree of under-fitting of the
+    CBVs to the light curve. It does so by measuring the mean residual target to
+    target Pearson correlation between the target under study and a selection of
+    neighboring SPOC SAP target light curves.
+
+    This function will begin with the given search_radius in arcseconds and
+    finds all neighboring targets. If not enough were found (< min_targets)
+    the search_radius is increased until a minimum number are found.
+
+    For TESS, the default search_radius is 5000 arcseconds.
+    For Kepler/K2, the default search_radius is 1000 arcseconds
+
+    The downloaded neighboring targets will normally be "aligned" to the
+    corrected_lc meaning the cadence numbers are used to align the targets
+    to the corrected_lc. However, if the CBVCorrector object was
+    instantiated with interpolated_cbvs=True then the targets will be
+    interpolated to the corrected_lc cadence times.
+
+    The returned under-fitting goodness metric is callibrated such that a
+    value of 0.95 means the residual correlations in the target is
+    equivalent to chance correlations of White Gaussian Noise.
+
+    Parameters
+    ----------
+    corrected_lc : LightCurve
+        Light curve from which systematics have been removed.
+    radius : float
+        Search radius to find neighboring targets in arcseconds
+    min_targets : int
+        Minimum number of targets to use in correlation metric
+        Using too few can cause unreliable results. Default = 30
+    max_targets : int
+        Maximum number of targets to use in correlation metric
+        Using too many can slow down the metric due to large data
+        download. Default = 50
+
+    Returns
+    -------
+    under_fitting_metric : float
+        A float in the range [0,1] where 0 => Bad, 1 => Good
+    """
+    # Download neighbor light curves
+    lcfCol = _download_neighbors(
+        lc=corrected_lc,
+        radius=radius,
+        min_targets=min_targets,
+        max_targets=max_targets,
+        flux_column="sap_flux",
+    )
```

Comment:
`_download_neighbors` is really slow to run.  I haven't completely dived through the code yet so am not certain if it is using the cache. But even if it is, there's a lot of functionality in there being called every time `underfit_metric_neighbors` is called and probably not viable in an optimizer loop to be run thousands of times. I think we need to have some type of cache for `underfit_metric_neighbors` so that `lc.search_neighbors` is only called once.

Reply:
I'm thinking the thing to do is inside the generic `Corrector` class have a cache, we then tell overfit_metric_lombscargle to either search for the neighborhood targets, or use a cache inside the `Corrector` object. This would be similar to what I had inside `CBVCorrector`,  just easier for non-cbvcorrector users to use.

Reply:
Thanks Jeff!  I confirm that `_download_neighbors` is very slow to run, not just on the first call, but also on every subsequent call.  That is unexpected.  It looks like the `download_all()` instruction isn't using the files cached on disk. My fault for not verifying that the caching is working correctly.  I'm going to investigate this bug.

Even when the caching of `download_all()` is fixed, I agree we will want to cache some of the additional alignment/interpolation work that takes place inside `underfit_metric_neighbors()`.  The cleanest way to do this will be to move the work that can be cached to a separate function (e.g. `_download_and_align_neighbors()`) and decorate that function with the `@memoization.cached` decorator which will automatically take care of caching for us.

Reply:
Confirmed caching code is functional and very fast.

Reply:
Awesome! 🎉 

Note: I merged #915 into master earlier today, which will speed up the initial call to `download_all()` before the caching provided within `metrics.py` kicks in.